pkg/reconciler/channel/channel.go

```diff
@@ -0,0 +1,412 @@
+			logging.FromContext(ctx).Infof("failed to record ready for Channel, %v", err)
+		}
+	}
+
+	return ch, err
+}
+
+func (c *Reconciler) syncSubscribers(ctx context.Context, channel *v1alpha1.Channel) error {
+	if channel.Status.Subscribers == nil {
+		channel.Status.Subscribers = []eventingduck.SubscriberStatus(nil)
+	}
+
+	subCreates := []eventingduck.SubscriberSpec(nil)
+	subUpdates := []eventingduck.SubscriberSpec(nil)
+	subDeletes := []eventingduck.SubscriberStatus(nil)
+
+	// Make a map of name to PullSubscription for lookup.
+	pullsubs := make(map[string]pubsubv1alpha1.PullSubscription)
+	if subs, err := c.getPullSubscriptions(ctx, channel); err != nil {
+		c.Logger.Infof("Failed to list PullSubscriptions, %s", err)
+	} else {
+		for _, s := range subs {
+			pullsubs[s.Name] = s
+		}
+	}
+
+	exists := make(map[types.UID]eventingduck.SubscriberStatus)
+	for _, s := range channel.Status.Subscribers {
+		exists[s.UID] = s
+	}
+
+	if channel.Spec.Subscribable != nil {
+		for _, want := range channel.Spec.Subscribable.Subscribers {
+			if got, ok := exists[want.UID]; !ok {
+				// If it does not exist, then update it.
+				subCreates = append(subCreates, want)
+			} else {
+				_, found := pullsubs[resources.GenerateSubscriptionName(want.UID)]
+				// If did not find or the PS has updated generation, update it.
+				if !found || got.ObservedGeneration != want.Generation {
+					subUpdates = append(subUpdates, want)
+				}
+			}
+			// Remove want from exists.
+			delete(exists, want.UID)
+		}
+	}
+
+	// Remaining exists will be deleted.
+	for _, e := range exists {
+		subDeletes = append(subDeletes, e)
+	}
+
+	for _, s := range subCreates {
+		genName := resources.GenerateSubscriptionName(s.UID)
+		c.Logger.Infof("Channel %q will create subscription %s", channel.Name, genName)
+
+		ps := resources.MakePullSubscription(&resources.PullSubscriptionArgs{
+			Owner:      channel,
+			Name:       genName,
+			Project:    channel.Spec.Project,
+			Topic:      channel.Status.TopicID,
+			Secret:     channel.Spec.Secret,
+			Labels:     resources.GetPullSubscriptionLabels(controllerAgentName, channel.Name, genName),
+			Subscriber: s,
+		})
+		ps, err := c.RunClientSet.PubsubV1alpha1().PullSubscriptions(channel.Namespace).Create(ps)
+		if err != nil {
+			c.Recorder.Eventf(channel, corev1.EventTypeWarning, "CreateSubscriberFailed", "Creating Subscriber %q failed", genName)
+			return err
+		}
+		c.Recorder.Eventf(channel, corev1.EventTypeNormal, "CreatedSubscriber", "Created Subscriber %q", genName)
+
+		channel.Status.Subscribers = append(channel.Status.Subscribers, eventingduck.SubscriberStatus{
+			UID:                s.UID,
+			ObservedGeneration: s.Generation,
+			// TODO: do I need the other fields?
+		})
+		return nil // Signal a re-reconcile.
+	}
+	for _, s := range subUpdates {
+		genName := resources.GenerateSubscriptionName(s.UID)
+		c.Logger.Infof("Channel %q will update subscription %s", channel.Name, genName)
```

Comment:
This is where I found some confusion, coupled with the lines below where we actually create it. They seem very similar to just create. Or put another way, why do you need subCreates and don't just throw everything into subUpdates. I think this is where the comments would help. It was difficult to visually diff what really was different about these two loops for the create case.

Reply:
https://github.com/GoogleCloudPlatform/cloud-run-events/issues/113